prologin/workernode/operations.py

```diff
@@ -18,336 +18,384 @@
-    result = await isolate_communicate(
-        cmd, limits, env=env, allowed_dirs=allowed_dirs, merge_outputs=True
-    )
-    return result.isolate_retcode, get_output(result)
-
 
-# TODO: refactor return value to a dataclass
-async def spawn_match(config, match_id, players, map_contents):
+class SpawnClient(Operation):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.result = {
+            'champion_id': None,
+            **self.result,
+        }
+        self.limits = {
+            'wall-time': self.config['isolate'].get('time_limit_secs', 350),
+            'mem': self.config['isolate'].get('mem_limit_MiB', 500) * 1000,
+            'processes': self.config['isolate'].get('processes', 50),
+            'fsize': 256,
+        }
+        self.isolate_allowed_dirs = [
+            '/var',
+            '/etc',
+            '/tmp',
+            os.path.dirname(self.config['path']['stechec_client']),
+            os.path.dirname(self.config['path']['rules']),
+        ]
+
+    def extract_champion(self, champion_tgz_b64: str) -> Path:
+        ctgz: bytes = b64decode(champion_tgz_b64)
+        untar(ctgz, self.isolator.path)
+        return Path('/box')
+
+    async def _run(
+        self,
+        *,
+        sockets_dir: os.PathLike,
+        req_addr: str,
+        sub_addr: str,
+        champion_id: int,
+        player_id: int,
+        champion_tgz_b64: str,
+        map_content: str = None,
+        order_id: int = None,
+    ):
+        self.result['champion_id'] = champion_id
+
+        self.isolate_allowed_dirs.append(str(sockets_dir) + ':rw')
+
+        champion_path = self.extract_champion(champion_tgz_b64)
+        env: Dict[str, str] = {
+            'CHAMPION_PATH': str(champion_path),
+            'HOME': '/tmp',
+        }
+
+        # fmt: off
+        cmd: List[str] = [
+            self.config['path']['stechec_client'],
+            "--name", str(player_id),
+            "--rules", self.config['path']['rules'],
+            "--champion", str(champion_path / 'champion.so'),
+            "--req_addr", req_addr,
+            "--sub_addr", sub_addr,
+            "--socket_timeout", "45000",
+            "--time", "1500",
+            "--verbose", "1",
+        ]
+        # fmt: on
+        if order_id is not None:
+            cmd += ["--client_id", str(order_id)]
+        if map_path := self.write_map(map_content):
+            cmd += ['--map', str(map_path)]
+
+        # Run the isolated client
+        await self.isolator.run(cmd, env=env, merge_outputs=True)
+
+
+async def compile_champion(config, champion_tgz_b64):
+    compile_champion = CompileChampion(config)
+    return await compile_champion(champion_tgz_b64=champion_tgz_b64)
+
+
+async def spawn_match(config, players, map_content):
     # Build the domain sockets
-    socket_dir = tempfile.TemporaryDirectory(prefix=f'workernode-{match_id}-')
+    socket_dir = tempfile.TemporaryDirectory(prefix='workernode-match-')
```

Comment:
Why remove the match_id? Without it its less straightforward to debug a stuck match.

Reply:
I did this to avoid having to pass the match ID in this function that doesn't really care about what the match ID is to compute it. I can revert that, although I doubt that we can debug anything in the socket directory.

Reply:
IIRC the trick was to have the match id somewhere in the command line in order to pivot to the box id, and then debug the container. Now that I think about it we can certainly have better debugging support than that, for example a status page available on HTTP that dumps the internals.

Reply:
Yeah, basically you could maintain the mapping between identifiers like match_id and other less important namespaces such as this temporary dir in a dedicated, isolated place. Fun fact, it's not pushed anywhere but my async rework of prolorpc contains a simple "debug" endpoint that just prints a dict and classes can easily register "observed" state (typically class attributes). Something similar would be welcome here.